### ISSUE TITLE
Cannot unslash before using variable in nonce check

### ISSUE THREAD

**Issue by rmccue** [WordPress-Coding-Standards/WordPress-Coding-Standards#572]
It's not possible to use a nonce check that uses an intermediate variable from `$_POST`. For example:

``` php
$args = wp_unslash( $_POST );
check_admin_referer( sprintf( 'wxr.import:%d', $args['import_id'] ) );
```

This means you have to duplicate the unslashing:

``` php
check_admin_referer( sprintf( 'wxr.import:%d', wp_unslash( $_POST['import_id'] ) ) );
$args = wp_unslash( $_POST );
```

This gets worse when you want to have the various `isset` checks for the variable, typecasting, etc. Seems like the check is too strict here, it should allow the nonce check later than it currently does.

**Comment by rmccue** [comment]
I am dumb, never mind.

**Comment by rmccue** [comment]
(Submitted comment on the wrong ticket; this one is still valid.)

**Comment by JDGrimes** [comment]
Assuming that `$_POST['import_id']` is an integer:

```php
if ( ! isset( $_POST['import_id'] ) ) {
	wp_die();
}

check_admin_referer( sprintf( 'wxr.import:%d', (int) $_POST['import_id'] ) );

$args = wp_unslash( $_POST );
```

But I understand that you really feel that there is a broader issue here. 

We do allow the use of `isset()` and sanitization (including casting) of the variable before the nonce check. The real issue here is the pattern of unslashing everything in `$_POST` without any sanitization. I think if you were performing sanitization simultaneously to that, it would work fine:

```php
$args = sanitize_import_args( wp_unslash( $_POST ) );
```

So yeah, the sniff just isn't designed to handle that, it is expecting you to unslash and sanitize each element of `$_POST` as you go along.

The reason that I made the sniff so strict about using the user input before the nonce check is because that could lead to CSRF. Anything that you do with the input data before the nonce check is subject to CSRF.

I'm not sure how we can make this better for you without also making the sniff more prone to false negatives. But perhaps we could introduce a flag that would disable some of the behavior?

In the mean time you can whitelist any problem spots like this:

```php
$args = wp_unslash( $_POST ); // WPCS: CSRF OK
check_admin_referer( sprintf( 'wxr.import:%d', $args['import_id'] ) );
```

**Comment by rmccue** [comment]
> The real issue here is the pattern of unslashing everything in `$_POST` without any sanitization.

I don't think that's really an "issue" (unless you mean in the technical sense). Unslashing is removing the transformation that WP applies to input vars, it doesn't really have any relation to sanitizing. I do get that there's technical issues with using an intermediary variable though.

> Anything that you do with the input data before the nonce check is subject to CSRF.

The sniffer can tell that `$args` isn't actually used for anything else, right?

The sniff currently has a whitelist for `isset`, my thinking is that `wp_unslash` should also be whitelisted.

**Comment by JDGrimes** [comment]
> I don't think that's really an "issue" (unless you mean in the technical sense).

I just meant that the sniff wasn't designed for that. It isn't necessarily an issue with your code. 😄 



> The sniffer can tell that $args isn't actually used for anything else, right?

Theoretically, yes. But this sniff doesn't actually check that. (PHPCS isn't the best tool for perform complex code-flow analysis [or really even fairly simple cases like this]. It can be done, but it can also make the sniff really complex.)



> The sniff currently has a whitelist for isset, my thinking is that wp_unslash should also be whitelisted.

Sounds reasonable.
